### ISSUE TITLE
Move to `sample_kbc` rather than `offline_fs_kbc` for nontee tests

### ISSUE THREAD

**Issue by fitzthum** [confidential-containers/confidential-containers#184]
We should stop thinking about the `offline_fs_kbc` as our non-tee KBC. Instead we should use the `sample_kbc` for this. The `offline_fs_kbc` is a KBC for platforms that protect the rootfs with encryption, but it isn't ideal for testing. I have mentioned this before, but here I will list the reasons more clearly.

**Why?**

First, the `offline_fs_kbc` makes it tricky for new users to test out encrypted images. If someone comes to the project without hardware, they will use the nontee runtime class. Currently the only way for them to use their own encrypted image with that runtime class is to modify the rootfs. This is not approachable for new users. 

[Here ](https://cloud-native.slack.com/archives/C039JSH0807/p1701806206433409) is an example thread where a new user tried to follow the quickstart guide, which explains how to provision resources via the KBS. Since they were using the nontee runtime class, this didn't work. With the sample attester, the flow for provisioning resources for TDX/SNP would be the same as for the non-tee runtime class.

As a side note, in the past we tried to get around a similar issue by creating the ssh-demo CRD. By switching to the sample-attester, we can probably get rid of this. 

Second, the `offline_fs_kbc` does not test the attestation code path at all. Currently our testing of the attestation code path is extremely spotty. In fact the KBS isn't explicitly referenced anywhere in the Kata or operator CI. The TDX CI does cover it, but this is up to the configuration of the CI node, which is behind a firewall. It would be better to have this crucial aspect of the project covered more directly by our CI. Of course the sample attester does not constitute a rigorous attestation, but it does trigger the complete attestation codepath from AA->KBS->AS and back.

Third, it could be misleading to expose flows that store secrets in the rootfs. Besides s390x, none of our platforms support encrypted root disks, so we risk confusing people when we write tests that do exactly that. Of course this is a less concrete concern than the previous two, but I think it's still significant.

Fourth, the `offline_fs_kbc` might not be compatible with advanced features of the CDH. As of today I think the CDH (like the `online_sev_kbc` does work with all of the new features that have been added to the CDH, but as we continue to add more advanced features, that might no longer be the case. If we want to test all these features in the non-tee CI, we'll probably need to use some kind of online KBC like the sample one.

**How?**

So how/when would we make the switch? I think writing tests for Kata main could be the right time. If we switch to the sample attester, provisioning resource for SNP/TDX and nontee would be very similar. I think this can help us writing the tests. Of course the s390x tests will still need to modify the initrd.

We can potentially keep the `offline_fs_kbc` in the rootfs for the nontee runtime class in case people do want to modify the initrd. With this approach we could potentially continue to support the existing approach as well as the sample attester in the nontee runtime class.

I am creating this issue today so that I can reference it in some small fixes to various components, but I am planning to pursue this more directly as well.

@ChengyuZhu6 @lindayu17 I am tagging you because you were working on some features for Kata that might involve attestation. Please ignore this entire post if it is not helpful, but if you have been wondering about how to setup tests that involve provisioning things via attestation, this might be useful. Basically I think you can write tests for the nontee runtime class by using the sample-kbc (which is included in the AA by default). Then we can add platform-specific tests as a second step.

@stevenhorsman @wainersm @fidencio @jepio @portersrc

**Comment by stevenhorsman** [comment]
Hey Tobin,

I'm not against us switching to sample_kbc if that makes more sense and is easier for users to set up for the non-TEE tests and just keep the offline_fs_kbc for the s390x tests, there are a couple of things I'd like your thoughts on, that could be done potentially as part of this wider issue, or done beforehand to prep for the kata-containers main testing:
1. This is almost certainly just a me issue, but do we have any documentation for how to use sample_kbc yet? I've had a look around and can't find anything, but am probably looking in the wrong place. Note, this isn't a reason to not switch over as our offline_fs_kbc docs are probably not perfect either, but we have some and I know how to use them and we have a fair few examples of usages to copy, whereas I've not needed to play with sample_kbc, so it would be good to see how straight forward it is to set-up and integrate into the kata tests
2. In image-rs I think that all of the current integration tests: https://github.com/confidential-containers/guest-components/tree/main/image-rs/tests use the offline_fs_kbc, so do you think we should add sample_kbc support there first to test it before it gets included in kata, or is it not straight-forward to get set-up in that sort of environment.

Thanks!

PS: I guess another thing to note (which again isn't a blocker), is that switching over to sample-kbc is likely to take a little bit more time, so might cause a slight delay to 0.9 features getting merged to main, but if we think this is the correct approach then I think it's worth any costs, but we'll want to decide sooner rather than later

**Comment by fitzthum** [comment]
> This is almost certainly just a me issue, but do we have any documentation for how to use sample_kbc yet?

Not really. I am doing some legwork to make sure it actually works and I can add some docs as part of that.

> In image-rs I think that all of the current integration tests: https://github.com/confidential-containers/guest-components/tree/main/image-rs/tests use the offline_fs_kbc, so do you think we should add sample_kbc support there first to test it before it gets included in kata, or is it not straight-forward to get set-up in that sort of environment.

I think it's fine to stick with the current approach for the image-rs integration tests. No need to go through the whole KBS roundtrip for that.

> I guess another thing to note (which again isn't a blocker), is that switching over to sample-kbc is likely to take a little bit more time, so might cause a slight delay to 0.9 features getting merged to main, but if we think this is the correct approach then I think it's worth any costs, but we'll want to decide sooner rather than later

I am hoping that this will make things easier for v0.9.0 but if it turns out to be a big can of worms we can just stick with the current approach.

**Comment by fitzthum** [comment]
Btw when I first wrote this I used the wrong term. We don't want the `sample_kbc`, we want the `cc_kbc` with the sample attester.

**Comment by portersrc** [comment]
I think this is a good idea mostly for reasons 1 and 2 (better for first-time users; and exercises attestation code path for testing). With respect to this:
> We can potentially keep the offline_fs_kbc in the rootfs for the nontee runtime class in case people do want to modify the initrd....
-- yes, I think it's better to keep that support there.

**Comment by fitzthum** [comment]
Ok, I think I have wrapped up the underlying work required to easily use the sample attester (https://github.com/confidential-containers/guest-components/pull/425 https://github.com/confidential-containers/guest-components/pull/426 https://github.com/confidential-containers/kbs/pull/283)

Now when we build the guest components, we can simply do `make build TEE_PLATFORM=none` and only the sample attester will be enabled. 

For the KBS, it's pretty to deploy with docker compose. There is a guide for this [here](https://github.com/confidential-containers/kbs/blob/main/kbs/docs/cluster.md). I will summarize here as well. 

Set up keys for the cluster.
```
openssl genpkey -algorithm ed25519 > kbs/config/private.key
openssl pkey -in kbs/config/private.key -pubout -out kbs/config/public.pub
```

Run cluster
```
docker compose up
```

Build the [client tool](https://github.com/confidential-containers/kbs/tree/main/kbs/tools/client) and provision a resource with something like
```
 ./kbs-client --url http://127.0.0.1:8080 config --auth-private-key ../../kbs/config/private.key  set-resource --path my_repo/resource_type/123abc --resource-file test_resource
```

By default the KBS will reject sample evidence, so you might also want to set a more permissive resource policy.
```
./kbs-client --url http://127.0.0.1:8080 config --auth-private-key ../../kbs/config/private.key  set-resource-policy --policy-file allow_all.rego
```

You can also set a policy that will require a particular launch measurement.

**Comment by wainersm** [comment]
Hi @fitzthum !

This sounds as a good idea.

Now that we have some code to install KBS for testing on Kata main and PR 8870 is in advanced state; I thought in start looking at porting the encryption tests.

Do you know if we can use the existing encrypted images?

**Comment by fitzthum** [comment]
Good question. Maybe we should actually make another issue for the encryted tests. Otherwise this one can probably be closed.

Regarding the images, we have [these ones](https://github.com/orgs/confidential-containers/packages/container/package/test-container) that should work. The singed images are broken and need to be regenerated but the encrypted ones are good. 

That said, I think we should consider making new images for the encrypted image tests where the encrypted image does not have identical content to the unencrypted image. There are some weird quirks with container caching that can cause confusion if you have an encrypted image and an unencrypted one that are identical. Under normal circumstances we should not hit these issues, but to be totally foolproof if might be good to have the encrypted image contain a file that says `ENCRYPTED` and for the test to check that this exists after decrypting the image.
